services/can_logger/can_logger.py

```diff
@@ -4,49 +4,69 @@
 import postgres
 import csv
 import os
+import time
+import sys
+import collections
 import multiprocessing as mp
+from datetime import datetime
+from psycopg2 import OperationalError
 
-def csv_init(host_interface):
-    
-    # Open the log file
-    
-    logfd = open('/data/log/' + host_interface + '.csv', mode = 'a') 
-    
-    # Make csv writer object using log file
-    
-    logcsv = csv.writer(logfd, delimiter = ',', quotechar = '"',
-                        quoting = csv.QUOTE_MINIMAL)
-    
-    return logcsv
 
-def write_to_csv(log, wr_buff):
-    
+def write_to_csv(wr_buff, can_bus):
+
     # Write to CSV from write buffer. First item is timestamp, second is
     # can_id and third is can_data
+    with open(f'/data/log/{can_bus}.csv', mode='a') as logfd:
 
-    for row in wr_buff:
-    
-        log.writerow(row[0], row[1], row[2])
+        logcsv = csv.writer(logfd, delimiter=',', quotechar='"',
+                            quoting=csv.QUOTE_MINIMAL)
+
+        logcsv.writerows(wr_buff)
+
+    print(f'{can_bus}: Wrote successfully to CSV')
+
+
+def write_to_db(db, wr_buff, can_bus):
+
+    # Convert from list into tuple and from tuple to string. Then remove
+    # the last comma from data.
 
-def write_to_db(db, wr_buff):
+    def format_data(data):
+
+        return str(data) + ','
+
+    data = ''.join(map(format_data, tuple(wr_buff)))
+    data = data.rstrip(',')
+
+    try:
+
+        db.run(
+            "INSERT INTO can(time, can_interface, can_id, can_data) VALUES {0};".format(data))
```

Comment:
I feel like there has to be a better way to do this.

`db.run("INSERT INTO can(time, can_interface, can_id, can_data) VALUES (%s, %s, %s, %s);", tuple(wr_buff))`
or 
`db.run("INSERT INTO can(time, can_interface, can_id, can_data) VALUES (%s, %s, %s, %s);", (str(wr_buff[0]), str(wr_buff[1]), str(wr_buff[2]), str(wr_buff[3])))` if all of wr_buff's types are not strings

If those don't work could you give me the error you are getting and an example of what `wr_buff` contains and the types?

Regardless the way it is now opens up SQL injection attacks should be changed to paramstyle before merging

Reply:
There might be, however the thing is that to insert multiple data points with one insert, I use the format_data function to convert the Python array into a single string representing a SQL array. Also, I already dropped `.format()` and instead used the recommended string formatting. This will be coming in a future commit after some testing.

Reply:
Really should be using parameterized queries so that escaping issues are properly dealt with. https://www.psycopg.org/docs/usage.html